File: src/models/relationships.py
"""Junction tables for relationships"""

from typing import Optional
from sqlalchemy import Column, String, ForeignKey, Integer, Enum as SQLEnum
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from enum import Enum
from .base import Base, IDMixin, TimestampMixin

class ActorRole(str, Enum):
    """Actor roles on Nouns"""
    OWNER = "owner"
    ASSIGNEE = "assignee"
    SME = "sme"
    RESOURCE = "resource"
    STAKEHOLDER = "stakeholder"
    AWARENESS = "awareness"

class NounActor(Base, IDMixin, TimestampMixin):
    """M:N relationship between Nouns and Actors with roles"""
    __tablename__ = "noun_actors"

    noun_id = Column(UUID(as_uuid=True), ForeignKey("nouns.id"), nullable=False, index=True)
    actor = Column(String(255), nullable=False)  # +person, @group, !vendor, *ai
    role = Column(SQLEnum(ActorRole), nullable=False)

class NounAssignment(Base, IDMixin, TimestampMixin):
    """Container assignments with Kanban sort order"""
    __tablename__ = "noun_assignments"

    noun_id = Column(UUID(as_uuid=True), ForeignKey("nouns.id"), nullable=False, index=True)
    container_id = Column(UUID(as_uuid=True), ForeignKey("nouns.id"), nullable=False, index=True)
    sort_order = Column(Integer, nullable=False, default=0)

class NounBlock(Base, IDMixin, TimestampMixin):
    """Blocker relationships"""
    __tablename__ = "noun_blocks"

    blocker_id = Column(UUID(as_uuid=True), ForeignKey("nouns.id"), nullable=False, index=True)
    target_id = Column(UUID(as_uuid=True), ForeignKey("nouns.id"), nullable=False, index=True)

class NounHashTag(Base, IDMixin, TimestampMixin):
    """Global tagging system"""
    __tablename__ = "noun_hashtags"

    noun_id = Column(UUID(as_uuid=True), ForeignKey("nouns.id"), nullable=False, index=True)
    tag = Column(String(255), nullable=False, index=True)

File: src/models/relationships.rs
//! Relationship models - junction tables

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

use super::ActorRole;

/// NounActor - M:N actor-role relationship on nouns
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct NounActor {
    pub id: Uuid,
    pub noun_id: Uuid,
    pub actor: String, // Actor string with prefix (+, @, !, *)
    pub role: ActorRole,
    pub created_at: DateTime<Utc>,
}

impl NounActor {
    pub fn new(noun_id: Uuid, actor: String, role: ActorRole) -> Self {
        Self {
            id: Uuid::new_v4(),
            noun_id,
            actor,
            role,
            created_at: Utc::now(),
        }
    }
}

/// NounAssignment - assigns nouns to containers (Group, Project, MileStone)
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct NounAssignment {
    pub id: Uuid,
    pub noun_id: Uuid,
    pub container_id: Uuid,
    pub sort_order: i32,
    pub created_at: DateTime<Utc>,
}

impl NounAssignment {
    pub fn new(noun_id: Uuid, container_id: Uuid, sort_order: i32) -> Self {
        Self {
            id: Uuid::new_v4(),
            noun_id,
            container_id,
            sort_order,
            created_at: Utc::now(),
        }
    }
}

/// NounBlock - blocker-target relationship
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct NounBlock {
    pub id: Uuid,
    pub blocker_id: Uuid, // Must be type=Blocker
    pub target_id: Uuid,
    pub reason: Option<String>,
    pub created_at: DateTime<Utc>,
}

impl NounBlock {
    pub fn new(blocker_id: Uuid, target_id: Uuid, reason: Option<String>) -> Self {
        Self {
            id: Uuid::new_v4(),
            blocker_id,
            target_id,
            reason,
            created_at: Utc::now(),
        }
    }
}

/// NounHashTag - global tagging
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct NounHashTag {
    pub id: Uuid,
    pub noun_id: Uuid,
    pub tag: String,
    pub created_at: DateTime<Utc>,
}

impl NounHashTag {
    pub fn new(noun_id: Uuid, tag: String) -> Self {
        Self {
            id: Uuid::new_v4(),
            noun_id,
            tag: tag.to_lowercase().trim_start_matches('#').to_string(),
            created_at: Utc::now(),
        }
    }
}

/// Request types
#[derive(Debug, Clone, Deserialize)]
pub struct AssignNounRequest {
    pub noun_id: Uuid,
    pub container_id: Uuid,
    #[serde(default)]
    pub sort_order: Option<i32>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct BlockNounRequest {
    pub blocker_id: Uuid,
    pub target_id: Uuid,
    #[serde(default)]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct AddActorRequest {
    pub actor: String,
    pub role: ActorRole,
}

#[derive(Debug, Clone, Deserialize)]
pub struct AddTagRequest {
    pub tag: String,
}
